src/qengine/opencl.cpp

```diff
@@ -233,24 +264,53 @@ void QEngineOCL::SetDevice(const int& dID, const bool& forceReInit)
 
     // create buffers on device (allocate space on GPU)
     if (didInit) {
-        complex* nStateVec = AllocStateVec(maxQPower);
-
-        oldQueue.enqueueMapBuffer(*stateBuffer, CL_TRUE, CL_MAP_READ, 0, sizeof(complex) * maxQPower, NULL);
-
-        std::copy(stateVec, stateVec + maxQPower, nStateVec);
-
-        cl::Event unmapEvent;
-        oldQueue.enqueueUnmapMemObject(*stateBuffer, stateVec, NULL, &unmapEvent);
-        unmapEvent.wait();
-
-        stateBuffer = std::make_shared<cl::Buffer>(
-            context, CL_MEM_USE_HOST_PTR | CL_MEM_READ_WRITE, sizeof(complex) * maxQPower, nStateVec);
-        free(stateVec);
-        stateVec = nStateVec;
+        // In this branch, the QEngineOCL was previously allocated, and now we need to copy its memory to a buffer
+        // that's accessible in a new device. (The old buffer is definitely not accessible to the new device.)
+
+        if (!stateVec) {
+            // We did not have host allocation, so we definitely have to copy device-local memory to host memory, then
+            // to a new device.
+            cl::CommandQueue nQueue = queue;
+            queue = oldQueue;
+
+            complex* nStateVec = AllocStateVec(maxQPower, true);
+            BufferPtr nStateBuffer = MakeStateVecBuffer(nStateVec);
+            cl::Event copyEvent;
+            oldQueue.enqueueCopyBuffer(
+                *stateBuffer, *nStateBuffer, 0, 0, sizeof(complex) * maxQPower, NULL, &copyEvent);
+            copyEvent.wait();
+
+            // Host RAM should now by synchronized.
+            queue = nQueue;
+            if (usingHostRam) {
+                // If we're using host RAM from here out, just create the buffer from the array pointer, in the context
+                // of the new device/queue.
+                stateBuffer = MakeStateVecBuffer(nStateVec);
+                stateVec = nStateVec;
+            } else {
+                // If we're not using host RAM from here, we need to copy into a device memory buffer.
+                stateBuffer = MakeStateVecBuffer(NULL);
+                queue.enqueueWriteBuffer(*stateBuffer, CL_TRUE, 0, sizeof(bitCapInt) * BCI_ARG_LEN, nStateVec);
+                free(nStateVec);
+            }
+        } else if (!usingHostRam) {
```

Comment:
Stay consistent with the sequence of tests. Above you have if(usingHostRam), and here you have if (!).